// flowgen/lib/printers/common.js
"use strict";

exports.__esModule = true;
exports.comment = exports.generics = exports.parseTypeReference = exports.methodSignature = exports.parameter = exports.typeParameter = exports.literalType = void 0;

var ts = _interopRequireWildcard(require("typescript"));

var _options = require("../options");

var _index = _interopRequireDefault(require("./index"));

var _env = require("../env");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

const literalType = node => {
  if (node.literal) {
    if (node.literal.kind === ts.SyntaxKind.StringLiteral) {
      return _index.default.node.printType(node.literal);
    } else if (node.literal.text) {
      return node.literal.text;
    } else {
      return _index.default.node.printType(node.literal);
    }
  }

  if (node.type.typeName) {
    return node.type.typeName.text;
  }

  return _index.default.node.printType(node.type);
};

exports.literalType = literalType;

const typeParameter = node => {
  let defaultType = "";
  let constraint = "";

  if (node.constraint) {
    constraint = `: ${_index.default.node.printType(node.constraint)}`;
  }

  if (!node.withoutDefault && node.default) {
    defaultType = `= ${_index.default.node.printType(node.default)}`;
  }

  return `${node.name.text}${constraint}${defaultType}`;
};

exports.typeParameter = typeParameter;

const parameter = param => {
  let left = "";

  if (param.modifiers) {
    for (const modifier of param.modifiers) {
      if (modifier.kind === ts.SyntaxKind.ReadonlyKeyword) left += "+";
    }
  }

  let right;

  if (param.name.kind === ts.SyntaxKind.ObjectBindingPattern || param.name.kind === ts.SyntaxKind.ArrayBindingPattern) {
    left = `x`;
  } else {
    left += _index.default.node.printType(param.name);
  }

  if (!param.type) {
    if (param.name.kind === ts.SyntaxKind.ObjectBindingPattern) {
      if (param.name.elements.length > 0) {
        right = `{${param.name.elements.map(element => `${_index.default.node.printType(element)}: any`).join(", ")}}`;
      } else {
        right = "{}";
      }
    } else {
      right = "any";
    }
  } else {
    right = _index.default.node.printType(param.type);
  }

  if (param.questionToken && param.name.kind !== ts.SyntaxKind.ComputedPropertyName) {
    left += "?";
  }

  if (param.questionToken && param.name.kind === ts.SyntaxKind.ComputedPropertyName) {
    right = `(${right}) | void`;
  }

  if (param.dotDotDotToken) {
    left = "..." + left;
  }

  return `${left}: ${right}`;
};

exports.parameter = parameter;

const methodSignature = param => {
  let left = "";
  let isMethod = true;

  if (param.modifiers) {
    for (const modifier of param.modifiers) {
      if (modifier.kind === ts.SyntaxKind.ReadonlyKeyword) {
        left += "+";
        isMethod = false;
      }
    }
  }

  left += _index.default.node.printType(param.name);
  let right;

  if (param.questionToken && param.name.kind !== ts.SyntaxKind.ComputedPropertyName) {
    left += "?";
    isMethod = false;
  }

  if (param.name.kind === ts.SyntaxKind.ComputedPropertyName) {
    isMethod = false;
  }

  right = _index.default.functions.functionType(param, isMethod);

  if (param.questionToken && param.name.kind === ts.SyntaxKind.ComputedPropertyName) {
    right = `(${right}) | void`;
  }

  return `${left}${isMethod ? "" : ": "}${right}`;
};

exports.methodSignature = methodSignature;

const parseTypeReference = node => {
  if (node.typeName.left && node.typeName.right) {
    return _index.default.node.printType(node.typeName) + generics(node.typeArguments);
  }

  return node.typeName.text + generics(node.typeArguments);
};

exports.parseTypeReference = parseTypeReference;

const generics = (types, map = node => node) => {
  if (types && typeof types.length !== "undefined") {
    return `<${types.map(map).map(_index.default.node.printType).join(", ")}>`;
  }

  return "";
};

exports.generics = generics;

const jsDocPrintTag = tag => {
  const typeNameValue = tag.typeExpression && tag.typeExpression.type;
  const parameterNameValue = tag.name || tag.preParameterName;
  const typeName = typeNameValue ? ` {${_index.default.node.printType(typeNameValue)}}` : "";
  const parameterName = parameterNameValue ? ` ${tag.isBracketed ? `[${_index.default.node.printType(parameterNameValue)}]` : _index.default.node.printType(parameterNameValue)}` : "";
  const comment = tag.comment ? ` ${tag.comment}`.replace(/\n/g, "\n * ") : "";

  if (typeNameValue && typeNameValue.kind === ts.SyntaxKind.JSDocTypeLiteral) {
    let output = `\n * @${tag.tagName.text}${typeName}${parameterName}${comment}`;

    for (const key in typeNameValue.jsDocPropertyTags) {
      output += jsDocPrintTag(typeNameValue.jsDocPropertyTags[key]);
    }

    return output;
  }

  if (tag && tag.kind === ts.SyntaxKind.JSDocCallbackTag) {
    const parameterName = parameterNameValue ? _index.default.node.printType(parameterNameValue) : "";
    let output = `\n * @${tag.tagName.text} ${parameterName}${tag.comment}`;

    for (const param of tag.typeExpression.parameters) {
      output += jsDocPrintTag(param);
    }

    if (typeNameValue) output += jsDocPrintTag(typeNameValue);
    return output;
  }

  return `\n * @${tag.tagName.text}${typeName}${parameterName}${comment}`;
};

const comment = (0, _env.withEnv)((env, jsdoc) => {
  if (!(0, _options.opts)().jsdoc) return "";
  const blocks = jsdoc.map(doc => {
    const comment = (doc.comment ? `\n${doc.comment}` : "").replace(/\n/g, "\n * ");
    env.tsdoc = true;
    const tags = (doc.tags || []).map(jsDocPrintTag);
    env.tsdoc = false;
    return `/**${comment + tags.join("")}\n */\n`;
  }).join("");
  return `\n${blocks}`;
});
exports.comment = comment;
