// flowgen/lib/printers/relationships.js
"use strict";

exports.__esModule = true;
exports.namespaceProp = exports.namespace = exports.importExportSpecifier = exports.exporter = exports.moduleExports = void 0;

var ts = _interopRequireWildcard(require("typescript"));

var _options = require("../options");

var _namespaceManager = _interopRequireDefault(require("../namespace-manager"));

var _index = _interopRequireDefault(require("./index"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

const moduleExports = node => {
  let name = _index.default.node.printType(node.expression);

  if (node.isExportEquals && (0, _options.opts)().moduleExports) {
    return `declare module.exports: typeof ${name}\n`;
  } else {
    return `declare export default typeof ${name}\n`;
  }
};

exports.moduleExports = moduleExports;

const exporter = node => {
  let str = "";

  if (node.modifiers && node.modifiers.some(modifier => modifier.kind === ts.SyntaxKind.ExportKeyword)) {
    str += "export ";
  }

  if (node.modifiers && node.modifiers.some(modifier => modifier.kind === ts.SyntaxKind.DefaultKeyword)) {
    str += "default ";
  }

  return str;
};

exports.exporter = exporter;

const importExportSpecifier = node => {
  if (node.propertyName) {
    return `${_index.default.node.printType(node.propertyName)} as ${_index.default.node.printType(node.name)}`;
  }

  return _index.default.node.printType(node.name);
}; // TODO: move import here
// export const imports = (node: ImportNode, moduleName: string): string => {
//   let str = "import type ";
//   if (node.default) {
//     str += node.default;
//     if (node.explicit.length) {
//       str += ", ";
//     }
//   }
//   if (node.explicit.length) {
//     str += `{ ${node.explicit.join(", ")} }`;
//   }
//   str += ` from '${moduleName}'`;
//   return str;
// };


exports.importExportSpecifier = importExportSpecifier;

const namespace = (name, hidePunctuation = false) => {
  if (_namespaceManager.default.nsExists(name)) {
    return `${name}${hidePunctuation ? "" : "$"}`;
  }

  return name + (hidePunctuation ? "" : ".");
};

exports.namespace = namespace;

const namespaceProp = (name, hidePunctuation = false) => {
  if (_namespaceManager.default.nsPropExists(name)) {
    return `${_namespaceManager.default.getNSForProp(name)}${hidePunctuation ? "" : "$"}${name}`;
  }

  return name;
};

exports.namespaceProp = namespaceProp;
