// flowgen/lib/nodes/module.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var logger = _interopRequireWildcard(require("../logger"));

var _node = _interopRequireDefault(require("./node"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Module extends _node.default {
  constructor(node, name) {
    super(node);

    _defineProperty(this, "name", void 0);

    this.name = name;
  }

  addChild(name, child) {
    child.module = this.name;
    this.children[name] = child;
  }

  addChildren(name, child) {
    child.module = this.name;

    if (!this.children[name]) {
      this.children[name] = child;
      return;
    }

    if (this.children[name]) {
      for (const key in child.children) {
        this.children[name].addChildren(key, child.children[key]);
      }

      return;
    }
  }

  print(namespace, module, depth = 0) {
    const children = this.getChildren().map(child => {
      return child.print(undefined, this.name, depth + 1);
    }).join("\n\t");
    const node = `
    declare module '${this.name}' {
      ${children}
    }
    `;

    if (module !== undefined && depth === 1) {
      logger.error(this.raw, {
        type: "UnsupportedNestedModule"
      });
      return `/* ${node} */\n`;
    }

    return `${node}\n`;
  }

}

exports.default = Module;
