// flowgen/lib/nodes/namespace.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var ts = _interopRequireWildcard(require("typescript"));

var _lodash = require("lodash");

var _property = _interopRequireDefault(require("./property"));

var _node = _interopRequireDefault(require("./node"));

var _namespaceManager = _interopRequireDefault(require("../namespace-manager"));

var _printers = _interopRequireDefault(require("../printers"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Namespace extends _node.default {
  constructor(_name, _functions, property) {
    super(null);

    _defineProperty(this, "name", void 0);

    _defineProperty(this, "functions", void 0);

    _defineProperty(this, "property", void 0);

    _defineProperty(this, "print", (namespace = "", mod = "root", depth) => {
      const children = (0, _lodash.uniqBy)((0, _lodash.orderBy)(this.getChildren(), [a => a.isValue], ["desc"]), child => child.name.text || child.name);
      const functions = children.filter(child => child.raw && child.raw.kind === ts.SyntaxKind.FunctionDeclaration);
      const variables = (0, _lodash.flatten)(children.filter(child => child.raw && child.raw.kind === ts.SyntaxKind.VariableStatement).map(child => child.raw.declarationList.declarations));
      const enums = children.filter(child => child.raw && child.raw.kind === ts.SyntaxKind.EnumDeclaration);
      const classes = children.filter(child => child.raw && child.raw.kind === ts.SyntaxKind.ClassDeclaration);
      const namespaces = children.filter(child => {
        return child.isValue;
      });
      let name = this.name;

      if (namespace) {
        name = namespace + "$" + this.name;
      }

      const childrenNode = `${this.getChildren().map(child => {
        return child.print(name, mod, depth);
      }).join("\n\n")}`;

      if (this.functions.length || functions.length || variables.length || namespaces.length || classes.length || enums.length) {
        let topLevel = "";
        const nsGroup = `
      declare var npm$namespace$${name}: {|
        ${this.functions.map(propNode => `${_printers.default.functions.functionType(propNode.raw, true)},`).join("\n")}
        ${functions.map(child => {
          return `${child.name}: typeof ${name}$${child.name},`;
        }).join("\n")}
        ${variables.map(child => {
          return `${child.name.text}: typeof ${name}$${child.name.text},`;
        }).join("\n")}
        ${enums.map(child => {
          return `${child.name}: typeof ${name}$${child.name},`;
        }).join("\n")}
        ${classes.map(child => {
          return `${child.name}: typeof ${name}$${child.name},`;
        }).join("\n")}
        ${namespaces.map(child => {
          return `${child.name}: typeof npm$namespace$${name}$${child.name},`;
        }).join("\n")}
      |}\n`;

        if (namespace === "") {
          topLevel = `declare var ${name}: typeof npm$namespace$${name};\n`;
        }

        return topLevel + nsGroup + childrenNode;
      }

      return childrenNode;
    });

    this.name = _name;
    this.functions = _functions || [];
    this.property = property;

    _namespaceManager.default.register(_name);
  }

  addChild(name, child) {
    child.namespace = this.name;
    child.isValue = child.getChildren().some(node => {
      return node instanceof Namespace || node.raw.kind === ts.SyntaxKind.VariableStatement || node.raw.kind === ts.SyntaxKind.ClassDeclaration || node.raw.kind === ts.SyntaxKind.FunctionDeclaration || node.raw.kind === ts.SyntaxKind.EnumDeclaration;
    });

    _namespaceManager.default.registerProp(this.name, child.name);

    this.children[name] = child;
  }

  addChildren(name, child) {
    child.namespace = this.name;
    child.isValue = child.getChildren().some(node => node instanceof Namespace || node.raw.kind === ts.SyntaxKind.VariableStatement || node.raw.kind === ts.SyntaxKind.ClassDeclaration || node.raw.kind === ts.SyntaxKind.FunctionDeclaration || node.raw.kind === ts.SyntaxKind.EnumDeclaration);

    _namespaceManager.default.registerProp(this.name, child.name);

    if (!this.children[name]) {
      this.children[name] = child;
      return;
    }

    if (this.children[name]) {
      for (const key in child.children) {
        this.children[name].addChildren(key, child.children[key]);
      }

      return;
    }
  }

}

exports.default = Namespace;
