// flowgen/lib/cli/compiler.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _typescript = require("typescript");

var _typescriptCompiler = _interopRequireDefault(require("typescript-compiler"));

var _namespaceManager = _interopRequireDefault(require("../namespace-manager"));

var _options = require("../options");

var _checker = require("../checker");

var logger = _interopRequireWildcard(require("../logger"));

var _env = require("../env");

var _transformers = require("../parse/transformers");

var _parse = require("../parse");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const compile = (0, _env.withEnv)((env, sourceFile) => {
  const rootNode = (0, _parse.recursiveWalkTree)(sourceFile);
  const output = rootNode.getChildren().map(child => {
    return child.print();
  }).join("");
  return output;
});

const reset = options => {
  (0, _options.resetOptions)();

  if (options) {
    (0, _options.assignOptions)(options);
  }

  _namespaceManager.default.reset();
};

const transformers = [(0, _transformers.legacyModules)(), (0, _transformers.importEqualsTransformer)()];
/**
 * Compiles typescript files
 */

var _default = {
  reset,
  compile: compile.withEnv({}),

  setChecker(typeChecker) {
    _checker.checker.current = typeChecker;
  },

  getTransformers() {
    return transformers;
  },

  compileTest: (path, target) => {
    _typescriptCompiler.default.compile(path, "--module commonjs -t ES6 --out " + target);
  },
  compileDefinitionString: (string, options) => {
    reset(options);
    const compilerOptions = {
      noLib: true,
      target: _typescript.ScriptTarget.Latest
    };
    const compilerHost = (0, _typescript.createCompilerHost)({}, true);
    const oldSourceFile = compilerHost.getSourceFile;

    compilerHost.getSourceFile = (file, languageVersion) => {
      if (file === "file.ts") {
        return (0, _typescript.transform)( //$todo Flow has problems when switching variables instead of literals
        (0, _typescript.createSourceFile)("/dev/null", string, languageVersion, true), transformers, compilerOptions).transformed[0];
      }

      return oldSourceFile(file, languageVersion);
    };

    const program = (0, _typescript.createProgram)(["file.ts"], compilerOptions, compilerHost);
    _checker.checker.current = program.getTypeChecker();
    const sourceFile = program.getSourceFile("file.ts");
    if (!sourceFile) return "";
    logger.setSourceFile(sourceFile);
    return compile.withEnv({})(sourceFile);
  },
  compileDefinitionFile: (path, options, mapSourceCode = a => a) => {
    reset(options);
    const compilerOptions = {
      noLib: true,
      target: _typescript.ScriptTarget.Latest
    };
    const compilerHost = (0, _typescript.createCompilerHost)({}, true);
    const oldSourceFile = compilerHost.getSourceFile;
    const oldReadFile = compilerHost.readFile;

    compilerHost.readFile = fileName => mapSourceCode(oldReadFile(fileName), fileName);

    compilerHost.getSourceFile = (file, languageVersion) => {
      if (file === path) {
        return (0, _typescript.transform)( //$todo Flow has problems when switching variables instead of literals
        (0, _typescript.createSourceFile)(file, compilerHost.readFile(file), languageVersion, true), transformers, compilerOptions).transformed[0];
      }

      return oldSourceFile(file, languageVersion);
    };

    const program = (0, _typescript.createProgram)([path], compilerOptions, compilerHost);
    _checker.checker.current = program.getTypeChecker();
    const sourceFile = program.getSourceFile(path);
    if (!sourceFile) return "";
    logger.setSourceFile(sourceFile);
    return compile.withEnv({})(sourceFile);
  },
  compileDefinitionFiles: (paths, options, mapSourceCode = a => a) => {
    const compilerOptions = {
      noLib: true,
      target: _typescript.ScriptTarget.Latest
    };
    const compilerHost = (0, _typescript.createCompilerHost)({}, true);
    const oldSourceFile = compilerHost.getSourceFile;
    const oldReadFile = compilerHost.readFile;

    compilerHost.readFile = fileName => mapSourceCode(oldReadFile(fileName), fileName);

    compilerHost.getSourceFile = (file, languageVersion) => {
      if (paths.includes(file)) {
        return (0, _typescript.transform)( //$todo Flow has problems when switching variables instead of literals
        (0, _typescript.createSourceFile)(file, compilerHost.readFile(file), languageVersion, true), transformers, compilerOptions).transformed[0];
      }

      return oldSourceFile(file, languageVersion);
    };

    const program = (0, _typescript.createProgram)(paths, compilerOptions, compilerHost);
    _checker.checker.current = program.getTypeChecker();
    return paths.map(path => {
      const sourceFile = program.getSourceFile(path);
      if (!sourceFile) return [path, ""];
      logger.setSourceFile(sourceFile);
      reset(options);
      return [path, compile.withEnv({})(sourceFile)];
    });
  }
};
exports.default = _default;
